### PR TITLE
Feature is bare metal servers

### PR DESCRIPTION
**Pull request by ujjwal-ibm**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
=== RUN   TestAccIBMISBMSProfilesDataSource_basic
--- PASS: TestAccIBMISBMSProfilesDataSource_basic (23.86s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 27.255s
```
```
=== RUN   TestAccIBMISBMSProfileDataSource_basic
--- PASS: TestAccIBMISBMSProfileDataSource_basic (47.90s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 50.396s
```
```
=== RUN   TestAccIBMISBMSDataSource_basic
--- PASS: TestAccIBMISBMSDataSource_basic (47.00s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 50.699s
```
```
=== RUN   TestAccIBMISBMSsDataSource_basic
--- PASS: TestAccIBMISBMSsDataSource_basic (619.84s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 622.958s
```
```
=== RUN   TestAccIBMISBareMetalServerDiskDataSource_basic
--- PASS: TestAccIBMISBareMetalServerDiskDataSource_basic (21.20s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 24.772s
```
```
=== RUN   TestAccIBMISBareMetalServerDisksDataSource_basic
--- PASS: TestAccIBMISBareMetalServerDisksDataSource_basic (19.92s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 23.369s
```
```
=== RUN   TestAccIBMISBareMetalServerInitializationDataSource_basic
--- PASS: TestAccIBMISBareMetalServerInitializationDataSource_basic (23.23s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 26.662s
```
```
=== RUN   TestAccIBMISBareMetalServerNetworkInterfacesDataSource_basic
--- PASS: TestAccIBMISBareMetalServerNetworkInterfacesDataSource_basic (26.22s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 30.809s
```
```
=== RUN   TestAccIBMISBareMetalServerNetworkInterfaceDataSource_basic
--- PASS: TestAccIBMISBareMetalServerNetworkInterfaceDataSource_basic (22.11s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 25.346s
```
```
=== RUN   TestAccIBMISBareMetalServerNetworkInterfaceFloatingIPsDataSource_basic
--- PASS: TestAccIBMISBareMetalServerNetworkInterfaceFloatingIPsDataSource_basic (22.89s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 26.006s
```
```
=== RUN   TestAccIBMISBareMetalServerNetworkInterfaceFloatingIPDataSource_basic
--- PASS: TestAccIBMISBareMetalServerNetworkInterfaceFloatingIPDataSource_basic (44.79s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 47.378s
```
```
=== RUN   TestAccIBMISBareMetalServerAction_basic
no message sent
--- PASS: TestAccIBMISBareMetalServerAction_basic (36.32s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     38.277s
```

### CODE REVIEW

File: website/docs/r/is_bare_metal_server_network_interface.markdown [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    name = "my-bms"
+    image = "r134-31c8ca90-2623-48d7-8cf7-737be6fc4c3e"
+    zone = "us-south-3"
+    keys = [ibm_is_ssh_key.sshkey.id]
+    primary_network_interface {
+      subnet     = ibm_is_subnet.subnet1.id
+    }
+    vpc = ibm_is_vpc.vpc1.id
+}
+resource ibm_is_bare_metal_server_network_interface this {
```
Comment (astha-jain) [inline website/docs/r/is_bare_metal_server_network_interface.markdown:31]: resource "ibm_is_bare_metal_server_network_interface" this {

File: ibm/data_source_ibm_is_bare_metal_server_initialization.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+							Type:        schema.TypeString,
+							Computed:    true,
+							Sensitive:   true,
+							Description: "The password at initialization, encrypted using encryption_key, and returned base64-encoded",
+						},
+						isBareMetalServerUserAccountPassword: {
+							Type:        schema.TypeString,
+							Computed:    true,
+							Sensitive:   true,
+							Description: "The password at initialization, encrypted using encryption_key, and returned base64-encoded",
```
Comment (deepaksibm) [inline ibm/data_source_ibm_is_bare_metal_server_initialization.go:95]: description is same as the encrypted password. give a different description

File: ibm/data_source_ibm_is_bare_metal_servers.go [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+										Type:     schema.TypeSet,
+										Computed: true,
+										Elem:     &schema.Schema{Type: schema.TypeString},
+										Set:      schema.HashString,
+									},
+									isBareMetalServerNicSubnet: {
+										Type:     schema.TypeString,
+										Computed: true,
+									},
+									isBareMetalServerNicPrimaryIP: {
```
Comment (deepaksibm) [inline ibm/data_source_ibm_is_bare_metal_servers.go:184]: If the response does not have primary_ip as in scoped spec, can we remove this attribute from schema and keep only primary_ipv4_address ?

File: ibm/data_source_ibm_is_bare_metal_server.go [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceIBMIsBareMetalServer() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceIBMISBareMetalServerRead,
+
+		Schema: map[string]*schema.Schema{
+			"identifier": {
```
Comment (deepaksibm) [inline ibm/data_source_ibm_is_bare_metal_server.go:22]: can we rename it to 'bare_metal_server' to be consistent with other datasources

File: ibm/data_source_ibm_is_bare_metal_server.go [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{isBareMetalServerName},
+				ValidateFunc:  InvokeDataSourceValidator("ibm_is_bare_metal_server", "identifier"),
+			},
+			isBareMetalServerName: {
```
Comment (deepaksibm) [inline ibm/data_source_ibm_is_bare_metal_server.go:28]: add atleastoneof for id and name

File: ibm/data_source_ibm_is_bare_metal_server_initialization.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+				password := base64.StdEncoding.EncodeToString(ciphertext)
+				if rsaKey != nil {
+					rng := rand.Reader
+					clearPassword, err := rsa.DecryptPKCS1v15(rng, rsaKey, ciphertext)
+					if err != nil {
+						return diag.FromErr(fmt.Errorf("Can not decrypt the password with the given key, %s", err))
+					}
+					password = string(clearPassword)
+				}
+				currAccount[isBareMetalServerUserAccountPassword] = password
```
Comment (deepaksibm) [inline ibm/data_source_ibm_is_bare_metal_server_initialization.go:228]: can we set this value only if private_key is provided and we are able to decrypt the password? or mention in the doc saying password will be same as encrypted_password if private_key not provided

File: ibm/data_source_ibm_is_bare_metal_server_initialization_test.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+			},
+		},
+	})
+}
+
+func testAccCheckIBMISBareMetalServerInitializationDataSourceConfig() string {
+	// status filter defaults to empty
+	return fmt.Sprintf(`
+      data "ibm_is_bare_metal_server_initialization" "test1" {
+		  bare_metal_server = ""
```
Comment (deepaksibm) [inline ibm/data_source_ibm_is_bare_metal_server_initialization_test.go:34]: provide a valid server id

File: ibm/data_source_ibm_is_bare_metal_server.go [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+		return diag.FromErr(err)
+	}
+	options := &vpcv1.GetBareMetalServerOptions{}
+	if id != "" {
+		options.ID = &id
+	}
+	// else if name != "" {
+	// 	options.Name = &name
+	// }
+	bms, response, err := sess.GetBareMetalServerWithContext(context, options)
```
Comment (deepaksibm) [inline ibm/data_source_ibm_is_bare_metal_server.go:415]: are we going to accept name as well ?

**Pull request closed without merging**